### PR TITLE
LTR-501, LTR-301, LTR-558 Series of Lite-On Light (ALS) and Proximity(PS) sensors

### PR DESCRIPTION
**Pull request by latonita**
# What does this implement/fix?

Component for ambient light and proximity sensors LTR-501, LTR-301, LTR-558

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** continuation of work on PR https://github.com/esphome/esphome/pull/6076

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3616

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [X] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
i2c:
# ...
sensor:
  - platform: ltr501
    address: 0x23
    type: ALS_PS
    ambient_light: "Ambient light"
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by kpfleming** [comment]
I fully support this PR getting merged, I've been using the ltr501 component from this branch for a few months now.

File: tests/components/ltr501/test.esp32.yaml
Comment (kbx81) [inline tests/components/ltr501/test.esp32.yaml:1]: This file (as well as the other arduino tests) will need to be renamed to be consistent with the updated test naming scheme.

Comment (latonita) [reply tests/components/ltr501/test.esp32.yaml:1]: I was going to ask 
1) whats the new schema
2) what to do with `No base test file [./tests/test_build_components/build_components_base.esp32-c3.yaml] for component test [./tests/components/ltr501/test.esp32-c3.yaml] found.`

Comment (latonita) [reply tests/components/ltr501/test.esp32.yaml:1]: oh, i see some files there in 'test-build...' 
I suppose my files shall have names like 'test.esp32-c3-ard.yaml' right?

Comment (latonita) [reply tests/components/ltr501/test.esp32.yaml:1]: @kbx81 done! build is passing now

**Comment by latonita** [comment]
@kbx81 do you think we might move forward with this request, already becoming quite old ? :) Thanks

File: esphome/components/ltr501/ltr_definitions_501.h [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+};
+
+// // PS Mode
+// enum PsMode : uint8_t {
+//   PS_MODE_STANDBY_00 = 0,  // default
+//   PS_MODE_STANDBY_01 = 1,
+//   PS_MODE_ACTIVE_10 = 2,
+//   PS_MODE_ACTIVE_11 = 3,
+// };
+
```
Comment (kbx81) [inline esphome/components/ltr501/ltr_definitions_501.h]: If this isn't used, let's just remove it.

File: esphome/components/ltr501/ltr501.cpp [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+
+  uint8_t tries = MAX_TRIES;
+  do {
+    ESP_LOGV(TAG, "Waiting chip to reset");
+    delay(2);
+    als_ctrl.raw = this->reg((uint8_t) CommandRegisters::ALS_CONTR).get();
+  } while (als_ctrl.sw_reset && tries--);  // while sw reset bit is on - keep waiting
+
+  if (als_ctrl.sw_reset) {
+    ESP_LOGW(TAG, "Failed to finalize reset procedure");
```
Comment (kbx81) [inline esphome/components/ltr501/ltr501.cpp]: So many words, haha
```suggestion
    ESP_LOGW(TAG, "Reset failed");
```

**Review by kbx81 — CHANGES_REQUESTED** [review]
Couple more nits, but otherwise LGTM

File: esphome/components/ltr501/ltr501.cpp (hunk 2 above)
Comment (latonita) [reply esphome/components/ltr501/ltr501.cpp]: This is funny ) thanks for the find. "when programmer wins over human"

File: esphome/components/ltr501/ltr_definitions_501.h (hunk 1 above)
Comment (latonita) [reply esphome/components/ltr501/ltr_definitions_501.h]: removed!

File: esphome/components/ltr501/ltr501.cpp [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+      if (this->is_ps_()) {
+        this->configure_ps_();
+      }
+
+      this->state_ = State::IDLE;
+      break;
+
+    case State::IDLE:
+      if (this->is_ps_()) {
+        check_and_trigger_ps_();
```
Comment (kbx81) [inline esphome/components/ltr501/ltr501.cpp]:
```suggestion
        this->check_and_trigger_ps_();
```

**Review by kbx81 — CHANGES_REQUESTED** [review]
One more

**Review by kbx81 — APPROVED** [review]

**Pull request merged**